### PR TITLE
Add supporting functionality and tests for OCSP_resp_find_status function

### PR DESCRIPTION
**Pull request by samuel40791765**
### Issues:
Addresses #CryptoAlg-362
Addresses #CryptoAlg-752

### Description of changes: 
s2n needs support OCSP stapling support from AWS-LC, if we wish to be used as s2n's cryptographic provider. We've decided to implement an OCSP patch in AWS-LC.
This pull request adds functionality and the supporting functions needed for OCSP response status finding. OCSP_cert_to_id extracts the |OCSP_CERTID| from a certificate chain for further use. OCSP_resp_find_status does checks on the OCSP response der with the  |OCSP_CERTID| extracted to see whether the certificate is "Good", "Revoked", or "UnKnown". 
- Design Doc: Under CryptoAlg-752 thread
- https://tools.ietf.org/html/rfc6960#section-2.2
- s2n's usage: https://github.com/aws/s2n-tls/blob/main/tls/s2n_x509_validator.c#L474-L483
- OpenSSL: 
    - https://github.com/openssl/openssl/blob/master/crypto/ocsp/ocsp_cl.c#L243-L300
    - https://github.com/openssl/openssl/blob/master/crypto/ocsp/ocsp_lib.c

### Call-outs:
- Normally, the response status checking is done after basic verification on the OCSP Response is done. OCSP verification functionality will be implemented in the next PR. 

### Testing:
 - Testing uses s2n's ocsp_response.der and ocsp_response_revoked.der test files from https://github.com/aws/s2n-tls/tree/main/tests/pems/ocsp
 - Testing Doc:  Under CryptoAlg-752 thread

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by dkostic — COMMENTED** [review]
First pass, I still have to review the test file. General comment: it'd be nice if more comments are added to the functions because it's difficult to understand what exactly is happening without knowing a fair bit about OCSP.

File: crypto/ocsp/ocsp_client.c [hunk 1]
```diff
@@ -30,3 +30,98 @@ OCSP_BASICRESP *OCSP_response_get1_basic(OCSP_RESPONSE *resp) {
...
+}
+
+int OCSP_resp_find(OCSP_BASICRESP *bs, OCSP_CERTID *id, int last)
+{
+  if (bs == NULL || id == NULL){
+    OPENSSL_PUT_ERROR(OCSP, ERR_R_PASSED_NULL_PARAMETER);
+    return -1;
+  }
+
+  STACK_OF(OCSP_SINGLERESP) *sresp = bs->tbsResponseData->responses;
```
Comment (torben-hansen) [inline crypto/ocsp/ocsp_client.c:49]: `tbsResponseData ` can be `NULL`?

File: crypto/ocsp/ocsp_client.c [hunk 2]
```diff
@@ -30,3 +30,98 @@ OCSP_BASICRESP *OCSP_response_get1_basic(OCSP_RESPONSE *resp) {
...
+  OCSP_SINGLERESP *single;
+
+  if (last < 0) {
+    last = 0;
+  } else {
+    last++;
+  }
+  for (size_t i = last; i < sk_OCSP_SINGLERESP_num(sresp); i++) {
+    single = sk_OCSP_SINGLERESP_value(sresp, i);
+    if (!OCSP_id_cmp(id, single->certId)) {
```
Comment (torben-hansen) [inline crypto/ocsp/ocsp_client.c:59]: `single` can be `NULL`?

File: crypto/ocsp/ocsp_client.c [hunk 3]
```diff
@@ -30,3 +30,98 @@ OCSP_BASICRESP *OCSP_response_get1_basic(OCSP_RESPONSE *resp) {
...
+                            ASN1_GENERALIZEDTIME **revtime,
+                            ASN1_GENERALIZEDTIME **thisupd,
+                            ASN1_GENERALIZEDTIME **nextupd) {
+  if (single == NULL) {
+    OPENSSL_PUT_ERROR(OCSP, ERR_R_PASSED_NULL_PARAMETER);
+    return -1;
+  }
+
+  OCSP_CERTSTATUS *cst = single->certStatus;
+  int status = cst->type;
```
Comment (torben-hansen) [inline crypto/ocsp/ocsp_client.c:76]: `cst` can be `NULL`?

File: crypto/ocsp/ocsp_client.c [hunk 4]
```diff
@@ -30,3 +30,98 @@ OCSP_BASICRESP *OCSP_response_get1_basic(OCSP_RESPONSE *resp) {
...
+  /*
+   * If certificate status is revoked, we look up certificate revocation
+   * time and reason
+   */
+  if (status == V_OCSP_CERTSTATUS_REVOKED) {
+    OCSP_REVOKEDINFO *rev = cst->value.revoked;
+    if (revtime != NULL) {
+      *revtime = rev->revocationTime;
```
Comment (torben-hansen) [inline crypto/ocsp/ocsp_client.c:84]: `rev` can be `NULL`?

File: crypto/ocsp/ocsp_lib.c [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+  X509_ALGOR *alg;
+  OCSP_CERTID *cid = NULL;
+  unsigned char md[EVP_MAX_MD_SIZE];
+
+  if ((cid = OCSP_CERTID_new()) == NULL) {
+    goto err;
+  }
+
+  alg = cid->hashAlgorithm;
+  ASN1_OBJECT_free(alg->algorithm);
```
Comment (torben-hansen) [inline crypto/ocsp/ocsp_lib.c:53]: `alg` can be `NULL`?

File: crypto/ocsp/ocsp_lib.c [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+int OCSP_id_issuer_cmp(const OCSP_CERTID *a, const OCSP_CERTID *b)
+{
+  if (a == NULL || b == NULL){
+    OPENSSL_PUT_ERROR(OCSP, ERR_R_PASSED_NULL_PARAMETER);
+    return -1;
+  }
+
+  int ret = OBJ_cmp(a->hashAlgorithm->algorithm, b->hashAlgorithm->algorithm);
```
Comment (torben-hansen) [inline crypto/ocsp/ocsp_lib.c:98]: `hashAlgorithm` can be `NULL`?

**Review by torben-hansen — COMMENTED** [review]
Style is to use `//` for code comments.

File: crypto/ocsp/ocsp_client.c (hunk 4 above)
Comment (samuel40791765) [reply crypto/ocsp/ocsp_client.c:84]: Was looking through RFC and noticed last paragraph in section 2.2 specified that when OCSP Responder is sending back the definition of a revoked status, it is required that the it must also specify with these fields.
https://datatracker.ietf.org/doc/html/rfc6960#section-2.2
Should we hard fail if NULL since it's supposed to be there, or should we assume that the responder has followed the RFC requirements as specified?

[the remark this answers is not in the document]
Comment (samuel40791765) [reply crypto/ocsp/ocsp_client.c:84]: Yeah, that makes a lot of sense. This pointer memory assumption was taken over from OpenSSL, but I also think it's safer to copy memory instead of pointing to the same pointer instead. I'll fix it for the returned time values.

File: crypto/ocsp/ocsp_lib.c (hunk 5 above)
Comment (samuel40791765) [reply crypto/ocsp/ocsp_lib.c:53]: The alg structure is created by the OCSP_CERTID_new ASN1 function above, so I assume it will be initialized for use already.

File: crypto/ocsp/ocsp_client.c (hunk 2 above)
Comment (samuel40791765) [reply crypto/ocsp/ocsp_client.c:59]: I assume that since we use the number of OCSP_SINGLERESP returned to loop through the stack, we won't run into a NULL here? (Stacks of OCSP_SINGLERESP are extracted from OCSP_BASICRESP, which are extracted directly from the OCSP_RESPONSE der)

File: crypto/ocsp/ocsp_lib.c [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    if (ASN1_STRING_copy(cid->serialNumber, serialNumber) == 0)
+      goto err;
+  }
+  return cid;
+  err:
+  OCSP_CERTID_free(cid);
+  return NULL;
+}
+
+int OCSP_id_issuer_cmp(const OCSP_CERTID *a, const OCSP_CERTID *b)
```
[the remark this answers is not in the document]
Comment (samuel40791765) [reply crypto/ocsp/ocsp_lib.c:91]: I also agree that the function names aren't the best. However, the OCSP_id_issuer_cmp function (and other OpenSSL Export OCSP functions) are part of the [OCSP outwards facing API used in OpenSSL](https://github.com/openssl/openssl/blob/OpenSSL_1_1_1-stable/include/openssl/ocsp.h#L232-L233). To avoid backwards compatibility problems, I think it would be best to keep them these function names the same for now.
We could change internal static function names for future imports though, if those seem to be unclear.

**Comment by samuel40791765** [comment]
Details for OCSP pipeline and function definitions are provided in link: 

- https://w.amazon.com/bin/view/AWSCryptoAlgorithms/Interns/sachiang/#HAWS-LCOCSPpatchphase

We are only looking to support the OCSP use case of s2n, which is parsing and verifying an OCSP RESPONSE. This means that only sections 2.2 and section 4.2 of the OCSP RFC will be the most relevant to this branch.  
- https://datatracker.ietf.org/doc/html/rfc6960#section-2.2
- https://datatracker.ietf.org/doc/html/rfc6960#section-4.2

File: crypto/ocsp/ocsp_client.c [hunk 8]
```diff
@@ -71,17 +81,19 @@ int OCSP_single_get0_status(OCSP_SINGLERESP *single, int *reason,
...
-   * If certificate status is revoked, we look up certificate revocation
-   * time and reason
-   */
+
+  // If certificate status is revoked, we look up certificate revocation time and reason
   if (status == V_OCSP_CERTSTATUS_REVOKED) {
     OCSP_REVOKEDINFO *rev = cst->value.revoked;
     if (revtime != NULL) {
-      *revtime = rev->revocationTime;
+      OPENSSL_memcpy(revtime, &rev->revocationTime, sizeof(&rev->revocationTime));
```
Comment (dkostic) [inline crypto/ocsp/ocsp_client.c:96]: I'm not sure what you want to achieve with this memcpy, it is equivalent to `*revtime = rev->revocationTime`. My question in the previous review relating to this is: if you do `*revtime = rev->revocationTime` you will now have 2 pointers pointing to the same piece of memory (`*revtime`  and `rev->revocationTime`) -- is this a potential problem once you exit the function, i.e., is there a chance that in another function you use `revtime` to change something in the data while you don't want it to be changed in `rev->revocationTime`. Another potential issue if we are not careful is double freeing...

File: crypto/ocsp/ocsp_client.c [hunk 9]
```diff
@@ -91,14 +103,12 @@ int OCSP_single_get0_status(OCSP_SINGLERESP *single, int *reason,
       }
     }
   }
-  /*
-   * Look up when certificate was last updated and when is next update time
-   */
+  // Look up when certificate was last updated and when is next update time
   if (thisupd != NULL) {
-    *thisupd = single->thisUpdate;
+    OPENSSL_memcpy(thisupd, &single->thisUpdate, sizeof(&single->thisUpdate));
```
Comment (dkostic) [inline crypto/ocsp/ocsp_client.c:108]: same comment as above.

File: crypto/ocsp/ocsp_client.c [hunk 10]
```diff
@@ -91,14 +103,12 @@ int OCSP_single_get0_status(OCSP_SINGLERESP *single, int *reason,
...
   }
   if (nextupd != NULL) {
-    *nextupd = single->nextUpdate;
+    OPENSSL_memcpy(nextupd, &single->nextUpdate, sizeof(&single->nextUpdate));
```
Comment (dkostic) [inline crypto/ocsp/ocsp_client.c:111]: same comment as above.

File: crypto/ocsp/ocsp_test.cc [hunk 11]
```diff
@@ -539,36 +539,36 @@ TEST(OCSPTest, TestGoodOCSP) {
...
+  // set up trust store and certificate chain
   bssl::UniquePtr<X509_STORE> trust_store(X509_STORE_new());
   X509_STORE_add_cert(trust_store.get(),LoadX509fromPEM(ca_cert).get());
   bssl::UniquePtr<STACK_OF(X509)> server_cert_chain = CertsToStack(
       {LoadX509fromPEM(server_cert).get(),LoadX509fromPEM(ca_cert).get()});;
 
   X509 *subject = sk_X509_value(server_cert_chain.get(), 0);
   X509 *issuer = nullptr;
-  /* find the issuer in the chain. If it's not there. Fail everything. */
+  // find the issuer in the certificate chain. If it's not there. Fail everything.
```
Comment (dkostic) [inline crypto/ocsp/ocsp_test.cc:550]: Nit:
```suggestion
  // Find the issuer in the certificate chain. If it's not there, fail everything.
```

File: crypto/ocsp/ocsp_test.cc [hunk 12]
```diff
@@ -539,36 +539,36 @@ TEST(OCSPTest, TestGoodOCSP) {
...
-
     if (issuer_value == X509_V_OK) {
       issuer = issuer_candidate;
       break;
     }
   }
   ASSERT_TRUE(issuer);
 
-  /* Expect basic verify here, but we skip the step for now because functionality has not been implemented yet */
+  // Expect basic verify here, but we skip the step for now because functionality has not been implemented yet
```
Comment (dkostic) [inline crypto/ocsp/ocsp_test.cc:562]: Nit, please try to use 80 character lines where possible.

File: crypto/ocsp/ocsp_client.c (hunk 8 above)
Comment (samuel40791765) [reply crypto/ocsp/ocsp_client.c:96]: Sorry, I didn't realize it was just the same pointer assignment. My original intention was to copy the values of `rev->revocationTime` over to `revtime` instead. 
That being said, the purpose of retrieving `revtime` is to extract the revocation time value from the der response sent back from the OCSP responser. The `revtime` value is extracted only to check if the timestamp is valid and is not intended to be changed in another function. 
I'll change it back to the original way for now.

File: crypto/ocsp/ocsp_client.c (hunk 9 above)
Comment (samuel40791765) [reply crypto/ocsp/ocsp_client.c:108]: Purpose of retrieving `thisupd` is to extract the last updated time value of the certificate from the der response sent back from the OCSP responser. The `thisupd` value is extracted only to check if the certificate timestamp is valid and is not intended to be changed in another function.

File: crypto/ocsp/ocsp_client.c (hunk 10 above)
Comment (samuel40791765) [reply crypto/ocsp/ocsp_client.c:111]: Purpose of retrieving `nextupd ` is to extract the time value of when the certificate should be updated again from the OCSP responder. The `nextupd ` value is extracted only to check if the certificate timestamp is valid and is not intended to be changed in another function.
Changing back for now.

**Review by dkostic — APPROVED** [review]

File: crypto/ocsp/ocsp_client.c [hunk 13]
```diff
@@ -30,3 +31,109 @@ OCSP_BASICRESP *OCSP_response_get1_basic(OCSP_RESPONSE *resp) {
...
+  // we look for index of equivalent |OCSP_CERTID| of issuer certificate in |OCSP_BASICRESP|
+  int single_idx = OCSP_resp_find(bs, id, -1);
+  if (single_idx < 0) {
+    return 0;
+  }
+  OCSP_SINGLERESP *single = OCSP_resp_get0(bs, single_idx);
+
+  // Extract the update time and revocation status of certificate sent back from OCSP responder
+  int single_status = OCSP_single_get0_status(single, reason, revtime, thisupd, nextupd);
+  if (status != NULL) {
```
Comment (torben-hansen) [inline crypto/ocsp/ocsp_client.c:135]: Should probably document this behaviour in the function description.

File: crypto/ocsp/ocsp_test.cc [hunk 14]
```diff
@@ -218,3 +525,101 @@ TEST(OCSPTest, TestBasic) {
...
+
+  // Set up trust store and certificate chain
+  bssl::UniquePtr<X509_STORE> trust_store(X509_STORE_new());
+  X509_STORE_add_cert(trust_store.get(),LoadX509fromPEM(ca_cert).get());
+  bssl::UniquePtr<STACK_OF(X509)> server_cert_chain = CertsToStack(
+      {LoadX509fromPEM(server_cert).get(),LoadX509fromPEM(ca_cert).get()});;
+
+  X509 *subject = sk_X509_value(server_cert_chain.get(), 0);
+  X509 *issuer = nullptr;
+  // Find the issuer in the certificate chain. If it's not there, fail everything.
```
Comment (torben-hansen) [inline crypto/ocsp/ocsp_test.cc:550]: Shouldn't we know this? Why do we have to look for it?

File: crypto/ocsp/ocsp_test.cc [hunk 15]
```diff
@@ -218,3 +525,101 @@ TEST(OCSPTest, TestBasic) {
...
+  // Convert issuer certificate to |OCSP_CERTID|
+  bssl::UniquePtr<OCSP_CERTID> cert_id = bssl::UniquePtr<OCSP_CERTID>(OCSP_cert_to_id(EVP_sha1(), subject, issuer));
+  ASSERT_TRUE(cert_id);
+
+  ASN1_GENERALIZEDTIME *revtime, *thisupd, *nextupd;
+  // Actual verification of the response
+  const int ocsp_resp_find_status_res = OCSP_resp_find_status(basic_response.get(), cert_id.get(), &status, &reason, &revtime, &thisupd, &nextupd);
+  ASSERT_EQ(1, ocsp_resp_find_status_res);
+  ASSERT_EQ(V_OCSP_CERTSTATUS_REVOKED, status);
+}
```
Comment (torben-hansen) [inline crypto/ocsp/ocsp_test.cc:625]: What about the third status code `V_OCSP_CERTSTATUS_UNKNOWN`?

File: crypto/ocsp/ocsp_test.cc [hunk 16]
```diff
@@ -4,13 +4,78 @@
 #include <gtest/gtest.h>
```
Comment (torben-hansen) [inline crypto/ocsp/ocsp_test.cc:4]: what other test cases are you adding?

File: crypto/ocsp/ocsp_test.cc [hunk 17]
```diff
@@ -218,3 +525,101 @@ TEST(OCSPTest, TestBasic) {
   basic_response = bssl::UniquePtr<OCSP_BASICRESP>(OCSP_response_get1_basic(ocsp_response.get()));
   ASSERT_TRUE(basic_response);
 }
+
+TEST(OCSPTest, TestGoodOCSP) {
```
Comment (torben-hansen) [inline crypto/ocsp/ocsp_test.cc:529]: You have some shared logic between the two good tests you have now. I envision more good tests covering more accept states. Could consider abstracting these tests further to make that easier.

**Review by torben-hansen — COMMENTED** [review]
Is there a particular reason for exporting all the functions in `ocsp.h`? From a quick count only 4/9 are used in s2n.

File: crypto/ocsp/ocsp_test.cc (hunk 15 above)
Comment (samuel40791765) [reply crypto/ocsp/ocsp_test.cc:625]: s2n has no tests in regards to the third status code `V_OCSP_CERTSTATUS_UNKNOWN`. I've already took note of this case and have added to the list of related tests to add after bare minimum functionality support for s2n is done. I'll provide a document list of tests that we should consider adding.

File: crypto/ocsp/ocsp_test.cc (hunk 16 above)
Comment (samuel40791765) [reply crypto/ocsp/ocsp_test.cc:4]: I'll provide a document list of tests that we should consider adding.

File: crypto/ocsp/ocsp_test.cc (hunk 14 above)
Comment (samuel40791765) [reply crypto/ocsp/ocsp_test.cc:550]: I think this is just an extra check to find the issuer in the cert chain, as s2n is doing here: https://github.com/aws/s2n-tls/blob/main/tls/s2n_x509_validator.c#L444-L456 
s2n also has unit tests that test against an invalid issuer, so I assumed that this is something we shouldn't always know by default. https://github.com/aws/s2n-tls/blob/main/tests/unit/s2n_x509_validator_test.c#L1078-L1124 
We could argue that this is outside the scope of OCSP and more of a certificate validation phase that s2n does on its own? If so, hard setting the issuer would make sense.

Comment (torben-hansen) [reply crypto/ocsp/ocsp_test.cc:550]: They have to support arbitrary chains. For this unit test, we don't.  Our chain is very simple:
```
  bssl::UniquePtr&lt;STACK_OF(X509)&gt; server_cert_chain = CertsToStack(      {LoadX509fromPEM(server_cert).get(),LoadX509fromPEM(ca_cert).get()});;
```
So, we can just skip this and index directly for what we need.

**Review by samuel40791765 — COMMENTED** [review]
> Is there a particular reason for exporting all the functions in `ocsp.h`? From a quick count only 4/9 are used in s2n.

Good call out. I originally exported all these functions externally just because OpenSSL had them in their API.
Meeting with @darylmartin100  and @andrewhop decided that we only export the main functions that s2n uses. This will minimize the amount of outwards facing APIs. We'll move the nested functions to `internal.h`.

File: crypto/ocsp/ocsp_lib.c [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+    OPENSSL_PUT_ERROR(OCSP, ERR_R_PASSED_NULL_PARAMETER);
+    return NULL;
+  }
+
+  const X509_NAME *iname;
+  const ASN1_INTEGER *serial;
+  ASN1_BIT_STRING *ikey;
+
+  if (dgst == NULL) {
+    dgst = EVP_sha1();
```
[the remark this answers is not in the document]
Comment (samuel40791765) [reply crypto/ocsp/ocsp_lib.c:18]: If we change this behavior, another place we'll need changing in future OCSP changes is within one of the `OCSP_basic_verify` internal static functions, where OSSL tries to find a `|OCSP_RESPID|` in a stack of certs using the keyhash. This function also defaults to sha1 as its digest algorithm.
https://github.com/openssl/openssl/blob/OpenSSL_1_1_1-stable/crypto/ocsp/ocsp_vfy.c#L170-L194

**Comment by samuel40791765** [comment]
Link for document discussing currently tested OCSP circumstances and OCSP tests to consider under CryptoAlg-752 thread

File: crypto/ocsp/ocsp_lib.c (hunk 18 above)
[the remark this answers is not in the document]
Comment (torben-hansen) [reply crypto/ocsp/ocsp_lib.c:18]: With respect to your last comment, that was implemented some 21 years ago.

When reading that issue, Matt points out that it is possible to parse the hash algorithm identifier and use that to create the correct |OCSP_RESPID| object. Hence, it is not impossible, yet not straightforward. With that, I would leave it as-is.

The only problem is then, that we don't have that particular function `OCSP_id_get0_info` which is needed :)

File: crypto/ocsp/ocsp_test.cc (hunk 17 above)
Comment (torben-hansen) [reply crypto/ocsp/ocsp_test.cc:529]: You know the expected status code for each test case. Pass that as part of the test case parameters.

Should be abstracted when adding more tests.

Comment (samuel40791765) [reply crypto/ocsp/ocsp_test.cc:536]: Abstraction added :)

File: crypto/ocsp/ocsp_test.cc [hunk 19]
```diff
@@ -199,22 +264,331 @@ static const uint8_t ocsp_response_der[] = {
...
+  X509 *subject = sk_X509_value(server_cert_chain.get(), 0);
+  X509 *issuer = sk_X509_value(server_cert_chain.get(), 1);
+
+  // Expect basic verify here, but we skip step for now since functionality has
+  // not been implemented yet.
+
+  int status = 0;
+  int reason = 0;
+  // Convert issuer certificate to |OCSP_CERTID|
+  bssl::UniquePtr<OCSP_CERTID> cert_id = bssl::UniquePtr<OCSP_CERTID>(OCSP_cert_to_id(EVP_sha1(), subject, issuer));
```
Comment (torben-hansen) [inline crypto/ocsp/ocsp_test.cc:556]: Considering the discussion about `sha1`. Maybe we should add a call with a `NULL` parameter to catch if the behaviour of `OCSP_cert_to_id` changes? Add a comment above the call to explain why it is added.

File: crypto/ocsp/ocsp_test.cc (hunk 17 above)
Comment (torben-hansen) [reply crypto/ocsp/ocsp_test.cc:536]: Nice nice :)

File: crypto/ocsp/ocsp_test.cc (hunk 19 above)
Comment (samuel40791765) [reply crypto/ocsp/ocsp_test.cc:556]: I added a second note in the function description for OCSP in `ocsp.h` to clarify this behavior and another test to see if the OCSP parsing behavior is effected when setting a `NULL` parameter in `OCSP_cert_to_id`.

**Review by torben-hansen — APPROVED** [review]
Nice
